[autograd] allow PyNode to persist error message

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33157 [autograd] enable graph level thread parallelism on CPU
* #35259 [autograd] add tests for simple reentrant and stackoverflow escape
* **#34845 [autograd] allow PyNode to persist error message**

This PR allows PyNode to persist the error message so that any pure C++
thread that runs autograd with custom Python autograd function can successfully
catpure the error message without maintaining a initial PyThreadState.

Differential Revision: [D20480685](https://our.internmc.facebook.com/intern/diff/D20480685)

wanchaol committed:
[autograd] allow PyNode to persist error message

This PR allows PyNode to persist the error message so that any pure C++
thread that runs autograd with custom Python autograd function can successfully
catpure the error message without maintaining a initial PyThreadState.

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -44,19 +44,24 @@ PyObject *THPFunctionClass = nullptr;
 namespace torch { namespace autograd {
 
+void PyNode::throw_python_error() {
+  auto python_err = python_error();
```
nit: auto&

wanchaol replied:
we are creating a new `python_error()` here, so it cann't be a lvalue reference.

jjlilley replied:
or just the conventional:
`    python_error err;
`

pritamdamania87 commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -44,19 +44,24 @@ PyObject *THPFunctionClass = nullptr;
 
```
Should we fix this too?

wanchaol replied:
we don't need to, for a similar reason https://github.com/pytorch/pytorch/pull/34845/files#r393387465

pritamdamania87 commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -104,7 +109,7 @@ auto PyNode::apply(variable_list&& inputs) -> variable_list {
   auto num_inputs = inputs.size();
   THPObjectPtr pyInputs(PyTuple_New(num_inputs));
-  if (!pyInputs) throw python_error();
+  if (!pyInputs) throw_python_error();
```
Looks like there are many more `throw python_error()` in this file. Should we replace those too?

wanchaol replied:
I only applied this change to PyNode, all other places of `throw python_error()`, they are actually functions that is used by the PyNode, so PyNode will catch those exceptions and persist them as well. We don't need to change the error throwing logic outside of the PyNode.

wanchaol committed:
Update on "[autograd] allow PyNode to persist error message"


This PR allows PyNode to persist the error message so that any pure C++
thread that runs autograd with custom Python autograd function can successfully
catpure the error message without maintaining a initial PyThreadState.

Differential Revision: [D20480685](https://our.internmc.facebook.com/intern/diff/D20480685)

[ghstack-poisoned]

jjlilley commented on torch/csrc/autograd/python_function.h:
```diff
@@ -27,6 +27,11 @@ struct PyNode : public Node {
   variable_list apply(variable_list&& inputs) override;
   variable_list legacy_apply(const variable_list& inputs);
 
+  // throw a persist python_error to allow it to be thrown
```
Maybe tweak the comment? 
 // Throw a python_error with the PyErr state persisted,
 // so that we don't lose the error state if the GIL is released,
 // or if the exception is passed to another thread.

jjlilley reviewed (approved):
Looks great, thanks for working on this.

albanD commented:
Hi,

I'm a bit confused here.
To run the PyNode, you need to have a valid python context.
Also PyNode already run properly on the different worker threads we have in the engine and the error are properly propagated back to the main thread.
Could you explain why this change is needed?

jjlilley commented:
@albanD The problem seems to be that releasing the GIL clears out the PyErr information.

We acquire the GIL inside this specific PyNode function, but since callers don't have the GIL, we will drop the error on exit.

So hence when we throw, we need to persist the PyErr state, otherwise, it's gone forever.

Previously, the codepaths were implicitly relying on the caller holding the GIL, so that the PyErr state didn't vanish before we persisted it. Now that the scope gil acquire is actually doing something for that function, we need to persist the error info right away.

albanD commented:
> Previously, the codepaths were implicitly relying on the caller holding the GIL

The worker threads in the current engine do not hold the GIL. So the caller to the PyNode in the current engine do not actually hold the GIL.

wanchaol commented:
> > Previously, the codepaths were implicitly relying on the caller holding the GIL
> 
> The worker threads in the current engine do not hold the GIL. So the caller to the PyNode in the current engine do not actually hold the GIL.

So in the current engine, we are creating a PyThreadState using `  pybind11::gil_scoped_acquire gil` when we start the CPU thread/worker in here  https://github.com/pytorch/pytorch/blob/325425ce9706012c413beec09aaf506e1cde86b1/torch/csrc/autograd/python_engine.cpp#L51

the code that acquiring gil will create PyThreadState is [here](https://github.com/pybind/pybind11/blob/master/include/pybind11/pybind11.h#L1912)

Since PyError store all its error indicators in the PyThreadState, and because we never go out of scope of `thread_init` in our current engine(since we call `thread_main` inside), the PyThreadState will remain there and all the error can be correctly captured. 

But with the new multithread autograd design, we remove the CPU thread and we can call autograd function (i.e. PyNode) in a pure C++ created thread (i.e. at::launch). So we did not have the PyThreadState created initially. When we get into `PyNode::apply`, it's the first time we acquire the GIL and create the PyThreadState, when we get out of the scope of `PyNode::apply`, Python will delete the PyThreadState if it found it's the last `gilstate_counter`, see (here)[https://github.com/pybind/pybind11/blob/master/include/pybind11/pybind11.h#L1959]. Therefore the PyError indicators all get lost and we could not retrieve the correct error message. That's the underlying rational for this change, we persist the error message early so that we could throw it correctly in a C++ thread as well.

albanD commented:
There is a very specific logic to make this work [here](https://github.com/pytorch/pytorch/blob/3e68d0c5d09a66bea3e4f5978f96d55c3b06c201/torch/csrc/autograd/python_engine.cpp#L48-L66).
Why does worker thread don't have a python state? That would avoid the creation/deletion of the state every time we execute something on the pool no?

wanchaol committed:
Update on "[autograd] allow PyNode to persist error message"


This PR allows PyNode to persist the error message so that any pure C++
thread that runs autograd with custom Python autograd function can successfully
catpure the error message without maintaining a initial PyThreadState.

Differential Revision: [D20480685](https://our.internmc.facebook.com/intern/diff/D20480685)

[ghstack-poisoned]

wanchaol commented:
> There is a very specific logic to make this work [here](https://github.com/pytorch/pytorch/blob/3e68d0c5d09a66bea3e4f5978f96d55c3b06c201/torch/csrc/autograd/python_engine.cpp#L48-L66).
> Why does worker thread don't have a python state? That would avoid the creation/deletion of the state every time we execute something on the pool no?

I think it's because the threadpool is implemented in C++ and we don't create a PyThreadState there, Python C API explanation [here](https://docs.python.org/2/c-api/init.html#non-python-created-threads). We don't want to create a PyThreadState for every `at::launch` calls, that would be super expensive and couldn't do that as it will make engine.cpp be python dependent. I think by just letting PyNode persist error message early is probably the solution we want to take.

albanD commented:
> We don't want to create a PyThreadState for every at::launch calls

No you want to create it once when the thread is created and added to the pool and that's it.

> couldn't do that as it will make engine.cpp be python dependent.

That is exactly why we have a PythonEngine.
The thread pool could have a similar mechanism that initialize the python state on each thread in the pool when they are created to speed up all following `at::launch` that will acquire the GIL (just like we do in the PythonEngine [here](https://github.com/pytorch/pytorch/blob/3e68d0c5d09a66bea3e4f5978f96d55c3b06c201/torch/csrc/autograd/python_engine.cpp#L48-L55)).
cc @ilia-cher  do you think this is something that is feasible?

ilia-cher commented:
Re. thread local variables: I'd suggest not to multiply entities without need and reuse a single thread local state when it is possible; I outlined my suggestion in https://github.com/pytorch/pytorch/issues/28520 and plan to send the PR this week.

wanchaol commented:
> Re. thread local variables: I'd suggest not to multiply entities without need and reuse a single thread local state when it is possible; I outlined my suggestion in #28520 and plan to send the PR this week.

@albanD @ilia-cher I agree with that, if our thread pool can initialize PyThreadState at the very initial stage when creating the threadpool, then that might solve the problem, otherwise if we do it in the `python_engine.cpp` today, we will basically re-creating the PyThreadState everytime we do `at::launch`. 

Right now I think this PR is a good alternate approach until that get supported.

ilia-cher commented:
if you need to run some init procedure in every thread pool's thread, once, and only at startup then existing api can be extended to cover this case fairly easily. I can prototype a PR that extends thread pool ctor interface and API to allow this.

ilia-cher commented:
> > Re. thread local variables: I'd suggest not to multiply entities without need and reuse a single thread local state when it is possible; I outlined my suggestion in #28520 and plan to send the PR this week.
> 
> @albanD @ilia-cher I agree with that, if our thread pool can initialize PyThreadState at the very initial stage when creating the threadpool, then that might solve the problem, otherwise if we do it in the `python_engine.cpp` today, we will basically re-creating the PyThreadState everytime we do `at::launch`.
> 
> Right now I think this PR is a good alternate approach until that get supported.

(my prev. comment on thread local variables was less relevant, as this concerns one time initialization procedure for threads)

wanchaol commented:
> if you need to run some init procedure in every thread pool's thread, once, and only at startup then existing api can be extended to cover this case fairly easily. I can prototype a PR that extends thread pool ctor interface and API to allow this.

@ilia-cher that would be nice! I am wondering when we should call this contructor as from what I know, the first time we call `at::launch`, we will create the pool, as it's also a singleton, will you provide the API and help us initialize the singleton and pass the initialized func to each thread in that thread pool?

ilia-cher commented:
@wanchaol just like you can do `torch::set_num_interop_threads` some time at startup, before thread pool is created, we could also do `torch::add_interop_thread_init_function` some time at startup, before any thread pool is created (we create thread pools lazily)

ilia-cher commented:
would this API cover this case? this has to be called once, at startup, before any work is started

ilia-cher commented:
(we might relax this requirement later, similarly to relaxing set_num_interop_threads, it would require some thread pool work and is not very hi-pri atm)

albanD commented:
Yes that sounds good, we can then add this call when we initialize the python module [here](https://github.com/pytorch/pytorch/blob/c747f098464f992797d799e69161f0eec8abe612/torch/csrc/Module.cpp#L103) and that will benefit both the autograd engine and the jit.

wanchaol commented:
> would this API cover this case? this has to be called once, at startup, before any work is started

@ilia-cher  yep that would be awesome and should cover this use case!

wanchaol committed:
Update on "[autograd] allow PyNode to persist error message"


This PR allows PyNode to persist the error message so that any pure C++
thread that runs autograd with custom Python autograd function can successfully
catpure the error message without maintaining a initial PyThreadState.

Differential Revision: [D20480685](https://our.internmc.facebook.com/intern/diff/D20480685)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/autograd/python_function.h:
```diff
@@ -27,6 +27,16 @@ struct PyNode : public Node {
+  // TODO: This is a temporary approach to allow C++ thread to correctly
+  // capture Python Error in autograd, remove this when c10 thread pool
+  // allow to do one time initialization.
+  // see discussion in https://github.com/pytorch/pytorch/pull/34845
```
nit: Can we add a link to the gh issue for the follow up here as well?

wanchaol committed:
Update on "[autograd] allow PyNode to persist error message"


This PR allows PyNode to persist the error message so that any pure C++
thread that runs autograd with custom Python autograd function can successfully
catpure the error message without maintaining a initial PyThreadState.

Differential Revision: [D20480685](https://our.internmc.facebook.com/intern/diff/D20480685)

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] allow PyNode to persist error message"


This PR allows PyNode to persist the error message so that any pure C++
thread that runs autograd with custom Python autograd function can successfully
catpure the error message without maintaining a initial PyThreadState.

Differential Revision: [D20480685](https://our.internmc.facebook.com/intern/diff/D20480685)

[ghstack-poisoned]
